daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -266,10 +265,29 @@ object TransactionCoder {
     * @tparam Cid Contract id type
     * @return decoded GenNode
     */
-  def decodeVersionedNodes[Nid, Cid](
+  private[lf] def decodeVersionedNode[Nid, Cid](
       decodeNid: DecodeNid[Nid],
       decodeCid: ValueCoder.DecodeCid[Cid],
-      txVersion: TransactionVersion,
+      enclosingVersion: TransactionVersion,
+      protoNode: TransactionOuterClass.Node,
+  ): Either[DecodeError, (Nid, VersionedNode[Nid, Cid])] =
+    for {
+      version <- if (enclosingVersion precedes TransactionVersions.minNodeVersion) {
+        Right(enclosingVersion)
+      } else {
+        decodeVersion(protoNode.getVersion).filterOrElse(
+          nodeVersion => !(enclosingVersion precedes nodeVersion),
+          DecodeError(
+            s"A transaction of version $enclosingVersion cannot contain node of newer version (${protoNode.getVersion})")
+        )
```

Comment:
This made my brain explode. I think the combination of the strange meaning of `filter` and the negation caused this. 🤯
This is a bit lengthier but perhaps makes it clearer what we mean:
```suggestion
        decodeVersion(protoNode.getVersion) match {
          case Left(err) => Left(err)
          case Right(nodeVersion) =>
            if enclosingVersion precedes nodeVersion
              Left(DecodeError(
                s"A transaction of version $enclosingVersion cannot contain node of newer version (${protoNode.getVersion})"))
            else
              Right(nodeVersion)
        }
```